akka-stream/src/main/scala/akka/stream/impl/ActorRefBackpressureSinkStage.scala

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.util
+
+import akka.actor._
+import akka.dispatch.sysmsg.{ DeathWatchNotification, SystemMessage, Watch }
+import akka.stream.stage.GraphStageLogic.StageActorRef
+import akka.stream.{ Inlet, SinkShape, ActorMaterializer, Attributes }
+import akka.stream.Attributes.InputBuffer
+import akka.stream.stage._
+
+/**
+ * INTERNAL API
+ */
+private[akka] class ActorRefBackpressureSinkStage[In](ref: ActorRef, onInitMessage: Any,
+                                                      ackMessage: Any,
+                                                      onCompleteMessage: Any,
+                                                      onFailureMessage: (Throwable) ⇒ Any)
+  extends GraphStage[SinkShape[In]] {
+  val in: Inlet[In] = Inlet[In]("ActorRefBackpressureSink.in")
+  override val shape: SinkShape[In] = SinkShape(in)
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+      implicit var self: StageActorRef = _
+
+      val buffer: util.Deque[In] = new util.ArrayDeque[In]()
+      var acknowledgementReceived = false
+      var completeReceived = false
+
+      override def keepGoingAfterAllPortsClosed: Boolean = true
+
+      private val callback: AsyncCallback[Unit] = getAsyncCallback((_: Unit) ⇒ {
+        if (!buffer.isEmpty) sendData()
+        else acknowledgementReceived = true
+      })
```

Comment:
Is getting the async callback before preStart always safe now? I may be mis-remembering that it's not. // @rkuhn